src/service/middleware/retry/mod.rs

```diff
@@ -0,0 +1,55 @@
+use futures_util::future;
+use http::{Request, Response};
+use hyper::{Body, Error};
+use tower::retry::Policy;
+
+#[derive(Clone)]
+pub enum RetryConfig {
+    None,
+    Simple(usize),
+}
+
+impl Policy<Request<String>, Response<hyper::Body>, hyper::Error> for RetryConfig {
+    type Future = futures_util::future::Ready<Self>;
+
+    fn retry(
+        &self,
+        _req: &Request<String>,
+        result: Result<&Response<Body>, &Error>,
+    ) -> Option<Self::Future> {
+        match self {
+            RetryConfig::None => None,
+            RetryConfig::Simple(count) => match result {
+                Ok(_) => None,
+                Err(_) => {
+                    if *count > 0 {
+                        Some(future::ready(RetryConfig::Simple(count - 1)))
+                    } else {
+                        None
+                    }
+                }
```

Comment:
Generally retries should only occur when the response status code is 5xx, 429, and (in the odd case of Github) 403s (because they chose to use that code to indicate rate-limit issues instead of 429s). Other 4xx codes generally mean the client/user did something wrong and sending the same exact request won't solve the problem. Whereas with 5xxs (server error), waiting a bit and trying again might help. And obviously waiting and trying the same request can help for rate-limit issues.

We might want to hardcode that into the client _or_ provide a predicate for library users to provide so they can provide their own logic.

Reply:
Code above currently only retries when unexpected http requests come in, for instance, EOF, or some other HTTP issues. I am a little hesitant about 403s?

I just added logic to handle 5**, and 429.

BTW tower approach does allow clients to create more complex/opinionated retry logic if they want and replace our implementation.